server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -708,34 +740,44 @@ protected void doRun() {
                         totalMetrics.preIngest();
                         final int slot = i;
                         final Releasable ref = refs.acquire();
+                        DocumentParsingObserver documentParsingObserver = documentParsingObserverSupplier.get();
+                        final IngestDocument ingestDocument = newIngestDocument(indexRequest, documentParsingObserver);
                         // the document listener gives us three-way logic: a document can fail processing (1), or it can
                         // be successfully processed. a successfully processed document can be kept (2) or dropped (3).
-                        final ActionListener<Boolean> documentListener = ActionListener.runAfter(new ActionListener<>() {
-                            @Override
-                            public void onResponse(Boolean kept) {
-                                assert kept != null;
-                                if (kept == false) {
-                                    onDropped.accept(slot);
+                        final ActionListener<IngestPipelinesExecutionResult> documentListener = ActionListener.runAfter(
+                            new ActionListener<>() {
+                                @Override
+                                public void onResponse(IngestPipelinesExecutionResult result) {
+                                    assert result != null;
+                                    if (result.success) {
+                                        if (result.shouldKeep == false) {
+                                            onDropped.accept(slot);
+                                        }
+                                    } else {
+                                        // We were given a failure result in the onResponse method, so we must store the failure
+                                        // Recover the original document state, track a failed ingest, and pass it along
+                                        updateIndexRequestMetadata(indexRequest, ingestDocument.getOriginalMetadata());
```

Comment:
Given that this is the only callsite for `ingestDocument.getOriginalMetadata()`, what about pulling it off and duplicating it from the `ingestDocument` when it was just freshly constructed at line 744? That is, does the `IngestDocument` class need to learn about tracking its own `originalMetadata` or is that really a concern for this class to learn about interacting with `IngestDocument`s?

I imagine that would drop an `IngestDocument.java` changes from the PR, too, and obviate any need to additional testing of that class, etc.

Reply:
This was a good idea and is now done